test/test_torch.py

```diff
@@ -3776,12 +3775,31 @@ def test_trilu_indices(self):
         self.assertEqual(
             x.triu(0).nonzero().transpose(0, 1), torch.triu_indices(3, 3))
 
-    def test_triu(self):
-        x = torch.rand(SIZE, SIZE)
-        res1 = torch.triu(x)
-        res2 = torch.Tensor()
-        torch.triu(x, out=res2)
-        self.assertEqual(res1, res2, 0)
+    @staticmethod
+    def _test_triu_tril(self, cast):
+        def run_test(shape, cast, diagonal):
+            x = cast(torch.randn(*shape))
+            res1 = torch.triu(x, diagonal=diagonal)
+            res2 = cast(torch.Tensor())
+            torch.triu(x, diagonal=diagonal, out=res2)
+            self.assertEqual(res1, res2, 0)
+            res1 = torch.tril(x, diagonal=diagonal)
+            res2 = cast(torch.Tensor())
+            torch.tril(x, diagonal=diagonal, out=res2)
+            self.assertEqual(res1, res2, 0)
+
+            # check by adding
+            self.assertEqual(x, x.triu(diagonal) + x.tril(diagonal - 1))
+
+        diagonals = [-2, -1, 0, 1, 2]
+        shapes = [(3, 3), (5, 3, 3), (7, 5, 3, 3),  # square matrices
```

Comment:
Can you test the following (edge) cases as well:
- 0-dim tensor (size `[]`)
- Tensor with 0 in one dimension (size `[0, 3, 3]` and/or `[3, 0]`)
- Tensor with 1 in one dimension (size `[1, 3]`, `[3, 1]`)

Reply:
Thanks for the suggestions. These have been added to the suite as well.